### PR TITLE
[V1] Decouple GPU and TPU `InputBatch`

### PR DESCRIPTION
**Pull request by afeldman-nm**
## Purpose

Decouple TPU and GPU `InputBatch` implementations. This PR adds a new `tpu_input_batch.py`, which is identical to `gpu_input_batch.py` but with all but TPU-oriented logic stripped out. Rationale: (1) TPU code lags GPU code in terms of feature support, so it makes sense to allow GPU to have a separate input batch from TPU to support more advanced features; (2) This PR is a prerequisite for the new logitsprocs programming model introduced in #16728 , which makes `InputBatch` modifications that are incompatible with TPU. 

## Test Plan

Covered by existing GPU and TPU unit tests.

## Test Result

N/A

## (Optional) Documentation Update

N/A

### CODE REVIEW

File: vllm/v1/worker/tpu_input_batch.py [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+            dtype=torch.int64,
+            pin_memory=self.pin_memory,
+        )
+        prompt_token_ids = prompt_token_ids_cpu_tensor.numpy()
+        prompt_token_ids[:] = self.token_ids_cpu[:self.
+                                                 num_reqs, :max_prompt_len]
+        # Use the value of vocab_size as a pad since we don't have a
+        # token_id of this value.
+        for i in range(self.num_reqs):
+            prompt_token_ids[i, self.num_prompt_tokens[i]:] = self.vocab_size
```
[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/v1/worker/tpu_input_batch.py:517]: Non-issue, the token ids in the vocab range over [0,`self.vocab_size`) which inherently excludes `self.vocab_size`

File: vllm/v1/worker/tpu_input_batch.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+    def condense(self, empty_req_indices: list[int]) -> None:
+        num_reqs = self.num_reqs
+        if num_reqs == 0:
+            # The batched states are empty.
+            self._req_ids.clear()
+            self.req_output_token_ids.clear()
+            return
+
+        # NOTE(woosuk): This function assumes that the empty_req_indices
+        # is sorted in descending order.
```
[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/v1/worker/tpu_input_batch.py:430]: Thanks, just pushed a commit which adds clarifying docstrings to `InputBatch.condense()` in both `gpu_input_batch.py` and `tpu_input_batch.py`. Thoughts on this fix?

File: vllm/v1/worker/lora_model_runner_mixin.py [hunk 3]
```diff
@@ -15,7 +16,14 @@
...
 from vllm.model_executor.models import supports_lora, supports_multimodal
-from vllm.v1.worker.gpu_input_batch import InputBatch
+from vllm.platforms import current_platform
+
+if current_platform.device_name == "tpu":
+    from vllm.v1.worker.tpu_input_batch import InputBatch as TPUInputBatch
+else:
+    from vllm.v1.worker.gpu_input_batch import InputBatch as GPUInputBatch
+
+InputBatch = Union[TPUInputBatch, GPUInputBatch]
```
Comment (njhill) [inline vllm/v1/worker/lora_model_runner_mixin.py]: I think this should work:
```suggestion
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from vllm.v1.worker.tpu_input_batch import InputBatch as TPUInputBatch
    from vllm.v1.worker.gpu_input_batch import InputBatch as GPUInputBatch

    InputBatch = Union[TPUInputBatch, GPUInputBatch]
```

**Review by njhill — COMMENTED** [review]
Thanks @afeldman-nm LGTM.

> This PR adds a new `tpu_input_batch.py`, which is identical to `gpu_input_batch.py` but with all but TPU-oriented logic stripped out.

presumably you meant something like "but with non-TPU-supported logic stripped out"?

File: vllm/v1/worker/tpu_input_batch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/v1/worker/tpu_input_batch.py:430]: Come on gemini wake up

[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/v1/worker/tpu_input_batch.py:430]: @njhill 

image

File: vllm/v1/worker/lora_model_runner_mixin.py (hunk 3 above)
Comment (afeldman-nm) [reply vllm/v1/worker/lora_model_runner_mixin.py]: Done, thanks.

**Review by njhill — APPROVED** [review]
Thanks @afeldman-nm.

As discussed we can look at how to reconcile this again once the LogitsProcessor rework is done (and hopefully support LPs with TPU somehow).

**Comment by afeldman-nm** [comment]
> Thanks @afeldman-nm.
> 
> As discussed we can look at how to reconcile this again once the LogitsProcessor rework is done (and hopefully support LPs with TPU somehow).

Certainly. To be clear, vLLM's TPU logic currently supports certain hard-coded logits processors; the gap which needs to be filled is specifically to (1) support the new logitsproc programming model introduced in #16728 , and (2) support logitsproc extensibility. Basically, to have feature parity with GPU.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks for separating TPU and GPU on `InputBatch`!

**Review by aarnphm — APPROVED** [review]

**Comment by afeldman-nm** [comment]
Thanks for the reviews and merge @njhill @yaochengji @aarnphm
